Custom Hopper Inventories

### Expected behavior

Custom Hopper Inventories should be openable by a plugin

### Observed/Actual behavior

an error throws when opening inventory with InventoryType hopper

### Steps/models to reproduce

1. have a plugin that wants to open the hopper inventory
2. trigger the action, that the plugin opens the inventory
3. look into the console, there is an error (See Other)

### Plugin and Datapack List

Achievements, AuktionsHaus, BankSystem, BoosterSystem, CaseOpening, CB3Core, ChatSystem, ChestShop, Citizens, CityBuildCore, ClanSystem, dynmap, FastAsyncWorldEdit (W|
orldEdit), floodgate, GeyserHacks, GeyserSkinManager, GrieferGoldSystem, HeadSystem, Helpsystem, HolographicDisplays, ImageMaps, InventoryManagementApi, InventorySystem, JobSystem, LanguageApi, Lev|
elFurnaces, LogBlock, LuckPerms, MoneySystem, MultiWorld*, NavigationPlugin, OldCombatMechanics, OnlineTimeSystem, PaperCloudPlugin, Perks, Plan, PlayerCache, Plot2Dynmap, PlotSquared, ProtocolLib,|
 RestPluginApi, SpecialItems, SystemChat, TeleportationSystem, TicketSystem, TradeSystem, ViaBackwards, ViaVersion, VoteSystem

### Paper version

This server is running Paper version git-Paper-340 (MC: 1.19.3) (Implementing API version 1.19.3-R0.1-SNAPSHOT) (Git: aa9c200)
You are running the latest version
Previous version: git-Paper-339 (MC: 1.19.3)

### Other

the error:

org.bukkit.command.CommandException: Unhandled exception executing command 'colorfloor' in plugin SpecialItems v1.0-SNAPSHOT                                                                         |
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:47) ~[paper-api-1.19.3-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:155) ~[paper-api-1.19.3-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.craftbukkit.v1_19_R2.CraftServer.dispatchCommand(CraftServer.java:929) ~[paper-1.19.3.jar:git-Paper-339]
        at org.bukkit.craftbukkit.v1_19_R2.command.BukkitCommandWrapper.run(BukkitCommandWrapper.java:64) ~[paper-1.19.3.jar:git-Paper-339]
        at com.mojang.brigadier.CommandDispatcher.execute(CommandDispatcher.java:264) ~[paper-1.19.3.jar:?]
        at net.minecraft.commands.Commands.performCommand(Commands.java:313) ~[?:?]
        at net.minecraft.commands.Commands.performCommand(Commands.java:297) ~[?:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.performChatCommand(ServerGamePacketListenerImpl.java:2290) ~[?:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.lambda$handleChatCommand$20(ServerGamePacketListenerImpl.java:2252) ~[?:?]
        at net.minecraft.util.thread.BlockableEventLoop.lambda$submitAsync$0(BlockableEventLoop.java:59) ~[?:?]
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768) ~[?:?]
        at net.minecraft.server.TickTask.run(TickTask.java:18) ~[paper-1.19.3.jar:git-Paper-339]
        at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:153) ~[?:?]
        at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24) ~[?:?]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1341) ~[paper-1.19.3.jar:git-Paper-339]
        at net.minecraft.server.MinecraftServer.d(MinecraftServer.java:197) ~[paper-1.19.3.jar:git-Paper-339]
        at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:126) ~[?:?]
        at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1318) ~[paper-1.19.3.jar:git-Paper-339]
        at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1311) ~[paper-1.19.3.jar:git-Paper-339]
        at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:136) ~[?:?]
        at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:1289) ~[paper-1.19.3.jar:git-Paper-339]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1177) ~[paper-1.19.3.jar:git-Paper-339]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:316) ~[paper-1.19.3.jar:git-Paper-339]
        at java.lang.Thread.run(Thread.java:1589) ~[?:?]
Caused by: java.lang.ClassCastException: class org.bukkit.craftbukkit.v1_19_R2.block.CraftBlockState cannot be cast to class io.papermc.paper.block.LockableTileState (org.bukkit.craftbukkit.v1_19_R
2.block.CraftBlockState and io.papermc.paper.block.LockableTileState are in unnamed module of loader java.net.URLClassLoader @7aec35a)
        at net.minecraft.world.level.block.entity.BaseContainerBlockEntity.canUnlock(BaseContainerBlockEntity.java:85) ~[?:?]
        at net.minecraft.world.level.block.entity.BaseContainerBlockEntity.canOpen(BaseContainerBlockEntity.java:72) ~[?:?]
        at net.minecraft.world.level.block.entity.RandomizableContainerBlockEntity.canOpen(RandomizableContainerBlockEntity.java:162) ~[?:?]
        at net.minecraft.world.level.block.entity.RandomizableContainerBlockEntity.createMenu(RandomizableContainerBlockEntity.java:168) ~[?:?]
        at net.minecraft.server.level.ServerPlayer.openMenu(ServerPlayer.java:1541) ~[?:?]
        at org.bukkit.craftbukkit.v1_19_R2.entity.CraftHumanEntity.openInventory(CraftHumanEntity.java:337) ~[paper-1.19.3.jar:git-Paper-339]
        at de.tobi.specialitems.commands.ColorFloorCommand.onCommand(ColorFloorCommand.java:32) ~[SpecialItems-1.0-SNAPSHOT.jar:?]
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[paper-api-1.19.3-R0.1-SNAPSHOT.jar:?]
        ... 23 more

--- a/patches/server/0948-Add-BlockLockCheckEvent.patch
+++ b/patches/server/0948-Add-BlockLockCheckEvent.patch
@@ -5,7 +5,7 @@ Subject: [PATCH] Add BlockLockCheckEvent
 
 
 diff --git a/src/main/java/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
-index a782994e2e53f2c4212c2d59ce740ebf00a826b0..3444c1a2e7f1dd938e42cdf0668e43be273b9116 100644
+index a782994e2e53f2c4212c2d59ce740ebf00a826b0..3d37c9a57c01c5035770e20873a801bf2f591cc5 100644
 --- a/src/main/java/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
 +++ b/src/main/java/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
 @@ -69,17 +69,44 @@ public abstract class BaseContainerBlockEntity extends BlockEntity implements Co
@@ -22,7 +22,7 @@ index a782994e2e53f2c4212c2d59ce740ebf00a826b0..3444c1a2e7f1dd938e42cdf0668e43be
 +        return canUnlock(player, lock, containerName, null);
 +    }
 +    public static boolean canUnlock(Player player, LockCode lock, Component containerName, @Nullable BlockEntity blockEntity) {
-+        if (player instanceof net.minecraft.server.level.ServerPlayer serverPlayer && blockEntity != null && blockEntity.getLevel() != null) {
++        if (player instanceof net.minecraft.server.level.ServerPlayer serverPlayer && blockEntity != null && blockEntity.getLevel() != null && blockEntity.getLevel().getBlockEntity(blockEntity.getBlockPos()) == blockEntity) {
 +            final org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(blockEntity.getLevel(), blockEntity.getBlockPos());
 +            net.kyori.adventure.text.Component lockedMessage = net.kyori.adventure.text.Component.translatable("container.isLocked", io.papermc.paper.adventure.PaperAdventure.asAdventure(containerName));
 +            net.kyori.adventure.sound.Sound lockedSound = net.kyori.adventure.sound.Sound.sound(org.bukkit.Sound.BLOCK_CHEST_LOCKED, net.kyori.adventure.sound.Sound.Source.BLOCK, 1.0F, 1.0F);

